core/dimensions.cpp

```diff
@@ -180,15 +180,28 @@ int32_t Dimensions::index(const Dim dim) const {
 ///
 /// Throws if there is a mismatching dimension extent.
 Dimensions merge(const Dimensions &a, const Dimensions &b) {
-  auto out(a);
-  for (const auto dim : b.labels()) {
-    if (a.contains(dim)) {
+  Dimensions out;
+  auto it = b.labels().begin();
+  auto end = b.labels().end();
+  for (const auto dim : a.labels()) {
+    // add any labels appearing *before* dim
```

Comment:
@jl-wynen This is the relevant changed behavior: It tries to keep outer dims as outer dims, but otherwise favors order as given by `a`.